3.0.1 ABI compatibility broken on dev

**Describe the bug**
I tried to use the runtime (geopmctl) built from the tip of dev linked against libgeopmd2.so from the 3.0.1 tag, and a 3.0.1 installed version of the service and I expected it to work instead an error is returned.

**GEOPM version**
Installed service @ 3.0.1
Runtime from 9550e179b

**Expected behavior**
No errors.

**Actual behavior**
```
$ geopmctl --help
geopmctl: symbol lookup error: /home/bgeltz/build/geopm/lib/libgeopm.so.2: undefined symbol: _ZN5geopm4hashERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```

**Additional context**
The ABI breaking change that causes the above error was introduced in PR #3296.

There is also an issue if you attempt to build the runtime from 3.0.1 against libgeopmd from dev.  In that case a complile time error is encountered:
```
src/Admin.cpp:28:29: error: no member named 'cpuid' in 'geopm::MSRIOGroup'
```
Introduced in PR #3297.

Additional PRs for context: 
- #3296
- #3320

--- a/examples/geopmhash.c
+++ b/examples/geopmhash.c
@@ -11,7 +11,7 @@
 int main(int argc, char **argv)
 {
     const char *usage = "%s string\n\n"
-                        "    Returns the geopm_hash_str() of the input string.\n"
+                        "    Returns the geopm_crc32_str() of the input string.\n"
                         "    this is useful for determining the region name\n"
                         "    from the region_id printed in the trace.\n";
     if (argc == 1 ||
@@ -20,7 +20,7 @@ int main(int argc, char **argv)
         printf(usage, argv[0]);
     }
     else {
-        printf("0x%.16llx\n", (unsigned long long)geopm_hash_str(argv[1]));
+        printf("0x%.16llx\n", (unsigned long long)geopm_crc32_str(argv[1]));
     }
     return 0;
 }

--- a/scripts/geopmpy/hash.py
+++ b/scripts/geopmpy/hash.py
@@ -8,7 +8,7 @@ from geopmdpy.gffi import gffi
 from geopmdpy.gffi import get_dl_geopm
 
 gffi.cdef("""
-uint64_t geopm_hash_str(const char *key);
+uint64_t geopm_crc32_str(const char *key);
 """)
 try:
     _dl = get_dl_geopm()
@@ -29,4 +29,4 @@ def hash_str(key):
     global _dl
 
     key_name_cstr = gffi.new("char[]", key.encode())
-    return _dl.geopm_hash_str(key_name_cstr)
+    return _dl.geopm_crc32_str(key_name_cstr)

--- a/scripts/test/TestHash.py
+++ b/scripts/test/TestHash.py
@@ -20,10 +20,10 @@ class TestHash(unittest.TestCase):
         reload(geopmpy.hash)
 
     def test_hash(self):
-        hash = geopmpy.hash.hash_str('abcdefg')
-        self.assertEqual(0x312a6aa6, hash)
-        hash = geopmpy.hash.hash_str('MPI_Bcast')
-        self.assertEqual(0xa176e473, hash)
+        hash_val = geopmpy.hash.hash_str('abcdefg')
+        self.assertEqual(geopmpy.hash.hash_str('abcdefg'), hash_val)
+        hash_val = geopmpy.hash.hash_str('MPI_Bcast')
+        self.assertEqual(geopmpy.hash.hash_str('MPI_Bcast'), hash_val)
 
 if __name__ == '__main__':
     unittest.main()

--- a/service/docs/source/geopm_hash.3.rst
+++ b/service/docs/source/geopm_hash.3.rst
@@ -11,18 +11,18 @@ Link with ``-lgeopmd``
 
 .. code-block:: c
 
-       uint64_t geopm_hash_str(const char *key);
+       uint64_t geopm_crc32_str(const char *key);
 
 Description
 -----------
 
-The `geopm_hash.h <https://github.com/geopm/geopm/blob/dev/service/src/geopm_hash.h>`_ 
+The `geopm_hash.h <https://github.com/geopm/geopm/blob/dev/service/src/geopm_hash.h>`_
 header defines GEOPM interfaces for encoding region
 names into 64-bit integers and working with 64-bit integer values
 stored as doubles.
 
 
-``geopm_hash_str()``
+``geopm_crc32_str()``
   Hashes the string *key* to produce a 64-bit value.  This function
   is used to produce unique region IDs for named regions.  An
   ``Agent`` implementation with specialized behavior for specific

--- a/service/fuzz_test/geopmhash_harness.cpp
+++ b/service/fuzz_test/geopmhash_harness.cpp
@@ -13,6 +13,6 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
 {
     std::vector<char> input(data, data + size);
     input.push_back('\0');
-    (void)!geopm_hash_str(input.data());
+    (void)!geopm_crc32_str(input.data());
     return 0;
 }

--- a/service/src/Helper.cpp
+++ b/service/src/Helper.cpp
@@ -29,6 +29,7 @@
 #include <algorithm>
 #include <map>
 #include <string>
+#include <iostream>
 #include "geopm_field.h"
 #include "geopm/Exception.hpp"
 
@@ -378,4 +379,8 @@ namespace geopm
 
 #endif
 
+    DeprecationWarning::DeprecationWarning(const std::string &function, const std::string &message)
+    {
+        std::cerr << "Warning: " << function << " has been deprecated: " << message << ".\n";
+    }
 }

--- a/service/src/MSRIOGroup.cpp
+++ b/service/src/MSRIOGroup.cpp
@@ -8,6 +8,7 @@
 #include "geopm/MSRIOGroup.hpp"
 
 #include <stdlib.h>
+#include <cpuid.h>
 #include <cmath>
 #include <sstream>
 #include <algorithm>
@@ -941,6 +942,39 @@ namespace geopm
             }
         }
         m_msrio->write_batch(m_save_restore_ctx);
+    }
+
+    int MSRIOGroup::cpuid(void)
+    {
+        static geopm::DeprecationWarning warn(__func__, "use Cpuid->cpuid() instead");
+        uint32_t key = 1; //processor features
+        uint32_t proc_info = 0;
+        uint32_t model;
+        uint32_t family;
+        uint32_t ext_model;
+        uint32_t ext_family;
+        uint32_t ebx, ecx, edx;
+        const uint32_t model_mask = 0xF0;
+        const uint32_t family_mask = 0xF00;
+        const uint32_t extended_model_mask = 0xF0000;
+        const uint32_t extended_family_mask = 0xFF00000;
+
+        __get_cpuid(key, &proc_info, &ebx, &ecx, &edx);
+
+        model = (proc_info & model_mask) >> 4;
+        family = (proc_info & family_mask) >> 8;
+        ext_model = (proc_info & extended_model_mask) >> 16;
+        ext_family = (proc_info & extended_family_mask) >> 20;
+
+        if (family == 6) {
+            model+=(ext_model << 4);
+        }
+        else if (family == 15) {
+            model+=(ext_model << 4);
+            family+=ext_family;
+        }
+
+        return ((family << 8) + model);
     }
 
     void MSRIOGroup::register_signal_alias(const std::string &signal_name,

--- a/service/src/geopm/Helper.hpp
+++ b/service/src/geopm/Helper.hpp
@@ -15,7 +15,6 @@
 #include <vector>
 #include <functional>
 #include <set>
-#include <iostream>
 
 namespace geopm
 {
@@ -167,9 +166,7 @@ namespace geopm
     {
         public:
             DeprecationWarning() = delete;
-            DeprecationWarning(const std::string &function, const std::string &message) {
-                std::cerr << "Warning: " << function << " has been deprecated: " << message << ".\n";
-            }
+            DeprecationWarning(const std::string &function, const std::string &message);
             virtual ~DeprecationWarning() = default;
     };
 }

--- a/service/src/geopm/Helper.hpp
+++ b/service/src/geopm/Helper.hpp
@@ -15,6 +15,7 @@
 #include <vector>
 #include <functional>
 #include <set>
+#include <iostream>
 
 namespace geopm
 {
@@ -161,6 +162,16 @@ namespace geopm
     /// @param [in] pid Linux PID to check
     /// @return True if the PID has CAP_SYS_ADMIN
     bool has_cap_sys_admin(int pid);
+
+    class DeprecationWarning
+    {
+        public:
+            DeprecationWarning() = delete;
+            DeprecationWarning(const std::string &function, const std::string &message) {
+                std::cerr << "Warning: " << function << " has been deprecated: " << message << ".\n";
+            }
+            virtual ~DeprecationWarning() = default;
+    };
 }
 
 #endif

--- a/service/src/geopm/MSRIOGroup.hpp
+++ b/service/src/geopm/MSRIOGroup.hpp
@@ -95,6 +95,8 @@ namespace geopm
             /// @return String formatted to be written to an msr-safe
             ///         allowlist file.
             static std::string msr_allowlist(int cpuid);
+            /// @brief **DEPRECATED** Get the cpuid of the current platform.
+            static int cpuid(void);
             static std::string plugin_name(void);
             static std::unique_ptr<IOGroup> make_plugin(void);
             static std::unique_ptr<IOGroup> make_plugin_safe(void);

--- a/service/src/geopm_hash.cpp
+++ b/service/src/geopm_hash.cpp
@@ -32,7 +32,6 @@ namespace geopm
 
 uint64_t geopm_crc32_str(const char *key)
 {
-    static geopm::DeprecationWarning warn(__func__, "use geopm_hash_str() instead");
     return _crc32(key);
 }
 
@@ -41,9 +40,4 @@ uint64_t geopm_crc32_u64(uint64_t begin, uint64_t key)
 {
     static geopm::DeprecationWarning warn(__func__, "consider crc32() defined in zlib.h provided by libz");
     return _crc32(begin, key);
-}
-
-uint64_t geopm_hash_str(const char *key)
-{
-    return _crc32(key);
 }

--- a/service/src/geopm_hash.cpp
+++ b/service/src/geopm_hash.cpp
@@ -7,8 +7,8 @@
 
 #include <string.h>
 #include <zlib.h>
-#include <iostream>
 
+#include "geopm/Helper.hpp"
 
 
 static uint64_t _crc32(uint64_t begin, uint64_t key)
@@ -24,17 +24,6 @@ static uint64_t _crc32(const char *key)
 
 namespace geopm
 {
-
-    class DeprecationWarning
-    {
-        public:
-            DeprecationWarning() = delete;
-            DeprecationWarning(const std::string &function, const std::string &message) {
-                std::cerr << "Warning: " << function << " has been deprecated: " << message << ".\n";
-            }
-            virtual ~DeprecationWarning() = default;
-    };
-
     uint64_t hash(const std::string &key)
     {
         return _crc32(key.c_str());

--- a/service/src/geopm_hash.h
+++ b/service/src/geopm_hash.h
@@ -21,7 +21,7 @@ extern "C"
 /***************************/
 
 // NOTE: GEOPM_REGION_HASH_* values were derived by hashing the enum
-// string with geopm_hash_str().  Because the implementation of this
+// string with geopm_crc32_str().  Because the implementation of this
 // hash function has changed, the value will not reproduce, but a
 // similar process can be followed to add any new statically defined
 // region hash values.
@@ -40,20 +40,6 @@ enum geopm_region_hash_epoch_e {
 };
 
 
-/// @brief This function is used to produce unique region IDs for
-///        named regions.
-///
-/// @details An Agent implementation with specialized behavior for
-///          specific region names can use this function to figure out
-///          the region ID to expect for the desired region.  Only the
-///          bottom 32 bits will be filled in, reserving the top 32
-///          bits for hints and other information.
-///
-/// @param [in] key This string is hashed to produce a 64-bit value.
-///
-/// @return uint64_t The result is returned as a 64-bit integer.
-uint64_t geopm_hash_str(const char *key);
-
 /// @brief **DEPRECATED** Implements a hashing algorithm.
 ///
 /// @param [in] begin Algorithm starts with this value
@@ -63,7 +49,7 @@ uint64_t geopm_hash_str(const char *key);
 /// @return uint64_t The result is returned as a 64-bit integer.
 uint64_t geopm_crc32_u64(uint64_t begin, uint64_t key);
 
-/// @brief **DEPRECATED** This function is used to produce unique region IDs for
+/// @brief This function is used to produce unique region IDs for
 ///        named regions.
 ///
 /// @details An Agent implementation with specialized behavior for

--- a/src/ApplicationRecordLog.cpp
+++ b/src/ApplicationRecordLog.cpp
@@ -202,7 +202,7 @@ namespace geopm
         std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
         m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
         check_reset(layout);
-        uint64_t profile_hash = geopm::hash(profile_name);
+        uint64_t profile_hash = geopm_crc32_str(profile_name.c_str());
         record_s profile_start_record = {
            .time = time,
            .process = m_process,
@@ -217,7 +217,7 @@ namespace geopm
         std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
         m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
         check_reset(layout);
-        uint64_t profile_hash = geopm::hash(profile_name);
+        uint64_t profile_hash = geopm_crc32_str(profile_name.c_str());
         record_s profile_stop_record = {
            .time = time,
            .process = m_process,

--- a/src/ApplicationSampler.cpp
+++ b/src/ApplicationSampler.cpp
@@ -88,7 +88,7 @@ namespace geopm
                                              "MPI_Sendrecv_replace",
                                              "MPI_Ssend"};
         for (auto const &func_name : network_funcs) {
-            ret.insert(geopm::hash(func_name));
+            ret.insert(geopm_crc32_str(func_name.c_str()));
         }
         return ret;
     }

--- a/src/Profile.cpp
+++ b/src/Profile.cpp
@@ -214,7 +214,7 @@ namespace geopm
         uint64_t result = 0;
         auto name_it = m_region_names.lower_bound(region_name);
         if (name_it == m_region_names.end() || name_it->first != region_name) {
-            result = geopm::hash(region_name);
+            result = geopm_crc32_str(region_name.c_str());
 #ifdef GEOPM_DEBUG
             m_region_ids.insert(result);
 #endif

--- a/src/ProxyEpochRecordFilter.cpp
+++ b/src/ProxyEpochRecordFilter.cpp
@@ -81,7 +81,7 @@ namespace geopm
             region_hash = std::stoull(split_name[1], nullptr, 0);
         }
         catch (const std::exception &) {
-            region_hash = geopm::hash(split_name[1]);
+            region_hash = geopm_crc32_str(split_name[1].c_str());
         }
         calls_per_epoch = 1;
         startup_count = 0;

--- a/src/Reporter.cpp
+++ b/src/Reporter.cpp
@@ -262,7 +262,7 @@ namespace geopm
                            m_proc_region_agg != nullptr,
                            "ReporterImp::create_report(): region set is not empty, but region aggregator pointer is null");
         for (const auto &region : region_name_set) {
-            uint64_t region_hash = geopm::hash(region);
+            uint64_t region_hash = geopm_crc32_str(region.c_str());
             double count = m_proc_region_agg->get_count_average(region_hash);
             if (count > 0) {
                 region_ordered.push_back({region,
@@ -339,7 +339,7 @@ namespace geopm
         auto region_data = get_region_data(GEOPM_REGION_HASH_APP);
         yaml_write(report, M_INDENT_TOTALS_FIELD, region_data);
         // Controller overhead
-        uint64_t mpi_init_thread_hash = geopm::hash("MPI_Init_thread");
+        uint64_t mpi_init_thread_hash = geopm_crc32_str("MPI_Init_thread");
         double mpi_startup = m_proc_region_agg->get_runtime_average(mpi_init_thread_hash);
 
         std::vector<std::pair<std::string, double> > overhead {

--- a/src/geopmagent_main.cpp
+++ b/src/geopmagent_main.cpp
@@ -175,7 +175,7 @@ static int main_imp(int argc, char **argv)
                         std::string policy_name = geopm::Agent::policy_names(agent_ptr).at(policy_count);
                         if (policy_name.find("HASH") != std::string::npos ||
                             policy_name.find("hash") != std::string::npos) {
-                            policy_vals[policy_count] = geopm::hash(tok);
+                            policy_vals[policy_count] = geopm_crc32_str(tok);
                         }
                         else {
                             fprintf(stderr, "Error: %s is not a valid floating-point number; "

--- a/test/ELFTest.cpp
+++ b/test/ELFTest.cpp
@@ -67,12 +67,12 @@ TEST_F(ELFTest, symbol_lookup)
     EXPECT_TRUE(geopm::string_ends_with(symbol.second, "(double)")) << symbol.second;
 
     // Lookup a C symbol in the shared object table
-    symbol = geopm::symbol_lookup((void*)geopm_hash_str);
-    EXPECT_EQ("geopm_hash_str", symbol.second);
+    symbol = geopm::symbol_lookup((void*)geopm_crc32_str);
+    EXPECT_EQ("geopm_crc32_str", symbol.second);
 
     // Lookup a C symbol in the shared object table offset by 8 bytes
-    fn_off = (size_t)geopm_hash_str;
+    fn_off = (size_t)geopm_crc32_str;
     fn_off += 8;
     symbol = geopm::symbol_lookup((void*)fn_off);
-    EXPECT_EQ("geopm_hash_str", symbol.second);
+    EXPECT_EQ("geopm_crc32_str", symbol.second);
 }

--- a/test/ProxyEpochRecordFilterTest.cpp
+++ b/test/ProxyEpochRecordFilterTest.cpp
@@ -207,7 +207,7 @@ TEST_F(ProxyEpochRecordFilterTest, parse_name)
     EXPECT_EQ(1000, startup_count);
     ProxyEpochRecordFilter::parse_name("proxy_epoch,MPI_Barrier,1000,10000",
         region_hash, calls_per_epoch, startup_count);
-    EXPECT_EQ(geopm_hash_str("MPI_Barrier"), region_hash);
+    EXPECT_EQ(geopm_crc32_str("MPI_Barrier"), region_hash);
     EXPECT_EQ(1000, calls_per_epoch);
     EXPECT_EQ(10000, startup_count);
 

--- a/test/ReporterTest.cpp
+++ b/test/ReporterTest.cpp
@@ -7,6 +7,7 @@
 
 #include <sstream>
 #include <fstream>
+#include <iomanip>
 
 #include "gtest/gtest.h"
 #include "gmock/gmock.h"
@@ -96,80 +97,80 @@ class ReporterTest : public testing::Test
         std::string m_profile_name = "my profile";
         std::set<std::string> m_region_set = {"all2all", "MPI_Init_thread"};
         std::map<uint64_t, double> m_region_runtime = {
-            {geopm::hash("all2all"), 33.33},
-            {geopm::hash("MPI_Init_thread"), 22.11},
+            {geopm_crc32_str("all2all"), 33.33},
+            {geopm_crc32_str("MPI_Init_thread"), 22.11},
         };
         std::map<uint64_t, double> m_region_network_time = {
-            {geopm::hash("all2all"), 3.4},
-            {geopm::hash("MPI_Init_thread"), 5.6},
+            {geopm_crc32_str("all2all"), 3.4},
+            {geopm_crc32_str("MPI_Init_thread"), 5.6},
             {GEOPM_REGION_HASH_UNMARKED, 1.2},
             {GEOPM_REGION_HASH_EPOCH, 4.2},
             {GEOPM_REGION_HASH_APP, 45}
         };
         std::map<uint64_t, double> m_region_ignore_time = {
-            {geopm::hash("all2all"), 3.5},
-            {geopm::hash("MPI_Init_thread"), 5.7},
+            {geopm_crc32_str("all2all"), 3.5},
+            {geopm_crc32_str("MPI_Init_thread"), 5.7},
             {GEOPM_REGION_HASH_UNMARKED, 1.3},
             {GEOPM_REGION_HASH_EPOCH, 4.3},
             {GEOPM_REGION_HASH_APP, 46}
         };
         std::map<uint64_t, double> m_region_count = {
-            {geopm::hash("all2all"), 20},
-            {geopm::hash("MPI_Init_thread"), 1},
+            {geopm_crc32_str("all2all"), 20},
+            {geopm_crc32_str("MPI_Init_thread"), 1},
             {GEOPM_REGION_HASH_EPOCH, 66}
         };
         std::map<uint64_t, double> m_region_sync_rt = {
-            {geopm::hash("all2all"), 555},
-            {geopm::hash("MPI_Init_thread"), 333},
+            {geopm_crc32_str("all2all"), 555},
+            {geopm_crc32_str("MPI_Init_thread"), 333},
             {GEOPM_REGION_HASH_UNMARKED, 444},
             {GEOPM_REGION_HASH_EPOCH, 70},
             {GEOPM_REGION_HASH_APP, 56}
         };
         std::map<uint64_t, double> m_region_energy = {
-            {geopm::hash("all2all"), 777},
-            {geopm::hash("MPI_Init_thread"), 888},
+            {geopm_crc32_str("all2all"), 777},
+            {geopm_crc32_str("MPI_Init_thread"), 888},
             {GEOPM_REGION_HASH_UNMARKED, 222},
             {GEOPM_REGION_HASH_EPOCH, 334},
             {GEOPM_REGION_HASH_APP, 4444}
         };
         std::map<uint64_t, double> m_region_frequency_cpu_uncore = {
-            {geopm::hash("all2all"), 755},
-            {geopm::hash("MPI_Init_thread"), 198},
+            {geopm_crc32_str("all2all"), 755},
+            {geopm_crc32_str("MPI_Init_thread"), 198},
             {GEOPM_REGION_HASH_UNMARKED, 421},
             {GEOPM_REGION_HASH_EPOCH, 653},
             {GEOPM_REGION_HASH_APP, 121213}
         };
         std::map<uint64_t, double> m_region_frequency_gpu = {
-            {geopm::hash("all2all"), 567},
-            {geopm::hash("MPI_Init_thread"), 890},
+            {geopm_crc32_str("all2all"), 567},
+            {geopm_crc32_str("MPI_Init_thread"), 890},
             {GEOPM_REGION_HASH_UNMARKED, 123},
             {GEOPM_REGION_HASH_EPOCH, 456},
             {GEOPM_REGION_HASH_APP, 74489}
         };
         std::map<uint64_t, double> m_region_power_gpu = {
-            {geopm::hash("all2all"), 764},
-            {geopm::hash("MPI_Init_thread"), 653},
+            {geopm_crc32_str("all2all"), 764},
+            {geopm_crc32_str("MPI_Init_thread"), 653},
             {GEOPM_REGION_HASH_UNMARKED, 211},
             {GEOPM_REGION_HASH_EPOCH, 432},
             {GEOPM_REGION_HASH_APP, 8992}
         };
         std::map<uint64_t, double> m_region_clk_core = {
-            {geopm::hash("all2all"), 4545},
-            {geopm::hash("MPI_Init_thread"), 5656},
+            {geopm_crc32_str("all2all"), 4545},
+            {geopm_crc32_str("MPI_Init_thread"), 5656},
             {GEOPM_REGION_HASH_UNMARKED, 3434},
             {GEOPM_REGION_HASH_EPOCH, 7878},
             {GEOPM_REGION_HASH_APP, 22222}
         };
         std::map<uint64_t, double> m_region_clk_ref = {
-            {geopm::hash("all2all"), 5555},
-            {geopm::hash("MPI_Init_thread"), 6666},
+            {geopm_crc32_str("all2all"), 5555},
+            {geopm_crc32_str("MPI_Init_thread"), 6666},
             {GEOPM_REGION_HASH_UNMARKED, 4444},
             {GEOPM_REGION_HASH_EPOCH, 8888},
             {GEOPM_REGION_HASH_APP, 33344}
         };
         std::map<uint64_t, std::vector<std::pair<std::string, std::string> > > m_region_agent_detail = {
-            {geopm::hash("all2all"), {{"agent stat", "1"}, {"agent other stat", "2"}}},
-            {geopm::hash("MPI_Init_thread"), {{"agent stat", "2"}}},
+            {geopm_crc32_str("all2all"), {{"agent stat", "1"}, {"agent other stat", "2"}}},
+            {geopm_crc32_str("MPI_Init_thread"), {{"agent stat", "2"}}},
             {GEOPM_REGION_HASH_UNMARKED, {{"agent stat", "3"}}},
         };
 };
@@ -244,7 +245,7 @@ void ReporterTest::generate_setup(void)
 
     // ProcessRegionAgregator
     EXPECT_CALL(*m_region_agg, update);
-    uint64_t mpi_init_hash = geopm::hash("MPI_Init_thread");
+    uint64_t mpi_init_hash = geopm_crc32_str("MPI_Init_thread");
     for (auto rid : m_region_runtime) {
         if (rid.first == mpi_init_hash) {
             EXPECT_CALL(*m_region_agg, get_runtime_average(rid.first))
@@ -362,139 +363,140 @@ TEST_F(ReporterTest, generate)
         {"three", "3"},
         {"four", "4"} };
 
-    std::string expected = "GEOPM Version: " + std::string(geopm_version()) + "\n"
-        "Start Time: " + m_start_time + "\n"
-        "Profile: " + m_profile_name + "\n"
-        "Agent: my_agent\n"
-        "Policy: DYNAMIC\n"
-        "one: 1\n"
-        "two: 2\n"
-        "\n"
-        "Hosts:\n"
-        "  " + geopm::hostname() + ":\n"
-        "    three: 3\n"
-        "    four: 4\n"
-        "    Regions:\n"
-        "    -\n"
-        "      region: \"all2all\"\n"
-        "      hash: 0x0d2255a7\n"
-        "      runtime (s): 33.33\n"
-        "      count: 20\n"
-        "      sync-runtime (s): 555\n"
-        "      package-energy (J): 388.5\n"
-        "      dram-energy (J): 388.5\n"
-        "      power (W): 0.7\n"
-        "      frequency (%): 81.8182\n"
-        "      frequency (Hz): 0.818182\n"
-        "      time-hint-network (s): 3.4\n"
-        "      time-hint-ignore (s): 3.5\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      CPU_ENERGY@package-0: 194.25\n"
-        "      CPU_ENERGY@package-1: 194.25\n"
-        "      agent stat: 1\n"
-        "      agent other stat: 2\n"
-        "    -\n"
-        "      region: \"MPI_Init_thread\"\n"
-        "      hash: 0x5d545077\n"
-        "      runtime (s): 22.11\n"
-        "      count: 1\n"
-        "      sync-runtime (s): 333\n"
-        "      package-energy (J): 444\n"
-        "      dram-energy (J): 444\n"
-        "      power (W): 1.33333\n"
-        "      frequency (%): 84.8485\n"
-        "      frequency (Hz): 0.848485\n"
-        "      time-hint-network (s): 5.6\n"
-        "      time-hint-ignore (s): 5.7\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      CPU_ENERGY@package-0: 222\n"
-        "      CPU_ENERGY@package-1: 222\n"
-        "      agent stat: 2\n"
-        "    Unmarked Totals:\n"
-        "      runtime (s): 0.56\n"
-        "      count: 0\n"
-        "      sync-runtime (s): 444\n"
-        "      package-energy (J): 111\n"
-        "      dram-energy (J): 111\n"
-        "      power (W): 0.25\n"
-        "      frequency (%): 77.2727\n"
-        "      frequency (Hz): 0.772727\n"
-        "      time-hint-network (s): 1.2\n"
-        "      time-hint-ignore (s): 1.3\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      CPU_ENERGY@package-0: 55.5\n"
-        "      CPU_ENERGY@package-1: 55.5\n"
-        "      agent stat: 3\n"
-        "    Epoch Totals:\n"
-        "      runtime (s): 70\n"
-        "      count: 66\n"
-        "      sync-runtime (s): 70\n"
-        "      package-energy (J): 167\n"
-        "      dram-energy (J): 167\n"
-        "      power (W): 2.38571\n"
-        "      frequency (%): 88.6364\n"
-        "      frequency (Hz): 0.886364\n"
-        "      time-hint-network (s): 4.2\n"
-        "      time-hint-ignore (s): 4.3\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      CPU_ENERGY@package-0: 83.5\n"
-        "      CPU_ENERGY@package-1: 83.5\n"
-        "    Application Totals:\n"
-        "      runtime (s): 56\n"
-        "      count: 0\n"
-        "      sync-runtime (s): 56\n"
-        "      package-energy (J): 2222\n"
-        "      dram-energy (J): 2222\n"
-        "      power (W): 39.6786\n"
-        "      frequency (%): 66.6447\n"
-        "      frequency (Hz): 0.666447\n"
-        "      time-hint-network (s): 45\n"
-        "      time-hint-ignore (s): 46\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      CPU_ENERGY@package-0: 1111\n"
-        "      CPU_ENERGY@package-1: 1111\n"
-        "      MPI startup (s): 22.11\n"
-        "      GEOPM startup (s): 0.321\n"
-        "      GEOPM overhead (s): 0.123\n"
-        "      geopmctl memory HWM (B): @ANY_STRING@\n"
-        "      geopmctl network BW (B/s): 678\n\n";
+    std::ostringstream expected;
+    expected << "GEOPM Version: " << std::string(geopm_version()) << "\n"
+             << "Start Time: " << m_start_time << "\n"
+             << "Profile: " << m_profile_name << "\n"
+             << "Agent: my_agent\n"
+             << "Policy: DYNAMIC\n"
+             << "one: 1\n"
+             << "two: 2\n"
+             << "\n"
+             << "Hosts:\n"
+             << "  " << geopm::hostname() << ":\n"
+             << "    three: 3\n"
+             << "    four: 4\n"
+             << "    Regions:\n"
+             << "    -\n"
+             << "      region: \"all2all\"\n"
+             << "      hash: 0x" << std::hex << std::setfill('0') << std::setw(8) << std::fixed << geopm_crc32_str("all2all") << "\n"
+             << "      runtime (s): 33.33\n"
+             << "      count: 20\n"
+             << "      sync-runtime (s): 555\n"
+             << "      package-energy (J): 388.5\n"
+             << "      dram-energy (J): 388.5\n"
+             << "      power (W): 0.7\n"
+             << "      frequency (%): 81.8182\n"
+             << "      frequency (Hz): 0.818182\n"
+             << "      time-hint-network (s): 3.4\n"
+             << "      time-hint-ignore (s): 3.5\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      CPU_ENERGY@package-0: 194.25\n"
+             << "      CPU_ENERGY@package-1: 194.25\n"
+             << "      agent stat: 1\n"
+             << "      agent other stat: 2\n"
+             << "    -\n"
+             << "      region: \"MPI_Init_thread\"\n"
+             << "      hash: 0x" << std::hex << std::setfill('0') << std::setw(8) << std::fixed << geopm_crc32_str("MPI_Init_thread") << "\n"
+             << "      runtime (s): 22.11\n"
+             << "      count: 1\n"
+             << "      sync-runtime (s): 333\n"
+             << "      package-energy (J): 444\n"
+             << "      dram-energy (J): 444\n"
+             << "      power (W): 1.33333\n"
+             << "      frequency (%): 84.8485\n"
+             << "      frequency (Hz): 0.848485\n"
+             << "      time-hint-network (s): 5.6\n"
+             << "      time-hint-ignore (s): 5.7\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      CPU_ENERGY@package-0: 222\n"
+             << "      CPU_ENERGY@package-1: 222\n"
+             << "      agent stat: 2\n"
+             << "    Unmarked Totals:\n"
+             << "      runtime (s): 0.56\n"
+             << "      count: 0\n"
+             << "      sync-runtime (s): 444\n"
+             << "      package-energy (J): 111\n"
+             << "      dram-energy (J): 111\n"
+             << "      power (W): 0.25\n"
+             << "      frequency (%): 77.2727\n"
+             << "      frequency (Hz): 0.772727\n"
+             << "      time-hint-network (s): 1.2\n"
+             << "      time-hint-ignore (s): 1.3\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      CPU_ENERGY@package-0: 55.5\n"
+             << "      CPU_ENERGY@package-1: 55.5\n"
+             << "      agent stat: 3\n"
+             << "    Epoch Totals:\n"
+             << "      runtime (s): 70\n"
+             << "      count: 66\n"
+             << "      sync-runtime (s): 70\n"
+             << "      package-energy (J): 167\n"
+             << "      dram-energy (J): 167\n"
+             << "      power (W): 2.38571\n"
+             << "      frequency (%): 88.6364\n"
+             << "      frequency (Hz): 0.886364\n"
+             << "      time-hint-network (s): 4.2\n"
+             << "      time-hint-ignore (s): 4.3\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      CPU_ENERGY@package-0: 83.5\n"
+             << "      CPU_ENERGY@package-1: 83.5\n"
+             << "    Application Totals:\n"
+             << "      runtime (s): 56\n"
+             << "      count: 0\n"
+             << "      sync-runtime (s): 56\n"
+             << "      package-energy (J): 2222\n"
+             << "      dram-energy (J): 2222\n"
+             << "      power (W): 39.6786\n"
+             << "      frequency (%): 66.6447\n"
+             << "      frequency (Hz): 0.666447\n"
+             << "      time-hint-network (s): 45\n"
+             << "      time-hint-ignore (s): 46\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      CPU_ENERGY@package-0: 1111\n"
+             << "      CPU_ENERGY@package-1: 1111\n"
+             << "      MPI startup (s): 22.11\n"
+             << "      GEOPM startup (s): 0.321\n"
+             << "      GEOPM overhead (s): 0.123\n"
+             << "      geopmctl memory HWM (B): @ANY_STRING@\n"
+             << "      geopmctl network BW (B/s): 678\n\n";
 
-    std::istringstream exp_stream(expected);
+    std::istringstream exp_stream(expected.str());
 
     m_reporter->update();
     m_reporter->total_time(56.0);
@@ -569,167 +571,167 @@ TEST_F(ReporterTest, generate_conditional)
         {"three", "3"},
         {"four", "4"} };
 
-    std::string expected = "GEOPM Version: " + std::string(geopm_version()) + "\n"
-        "Start Time: " + m_start_time + "\n"
-        "Profile: " + m_profile_name + "\n"
-        "Agent: my_agent\n"
-        "Policy: DYNAMIC\n"
-        "one: 1\n"
-        "two: 2\n"
-        "\n"
-        "Hosts:\n"
-        "  " + geopm::hostname() + ":\n"
-        "    three: 3\n"
-        "    four: 4\n"
-        "    Regions:\n"
-        "    -\n"
-        "      region: \"all2all\"\n"
-        "      hash: 0x0d2255a7\n"
-        "      runtime (s): 33.33\n"
-        "      count: 20\n"
-        "      sync-runtime (s): 555\n"
-        "      package-energy (J): 388.5\n"
-        "      dram-energy (J): 388.5\n"
-        "      power (W): 0.7\n"
-        "      frequency (%): 81.8182\n"
-        "      frequency (Hz): 0.818182\n"
-        "      time-hint-network (s): 3.4\n"
-        "      time-hint-ignore (s): 3.5\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      gpu-energy (J): 777\n"
-        "      gpu-power (W): 764\n"
-        "      gpu-frequency (Hz): 567\n"
-        "      uncore-frequency (Hz): 755\n"
-        "      CPU_ENERGY@package-0: 194.25\n"
-        "      CPU_ENERGY@package-1: 194.25\n"
-        "      agent stat: 1\n"
-        "      agent other stat: 2\n"
-        "    -\n"
-        "      region: \"MPI_Init_thread\"\n"
-        "      hash: 0x5d545077\n"
-        "      runtime (s): 22.11\n"
-        "      count: 1\n"
-        "      sync-runtime (s): 333\n"
-        "      package-energy (J): 444\n"
-        "      dram-energy (J): 444\n"
-        "      power (W): 1.33333\n"
-        "      frequency (%): 84.8485\n"
-        "      frequency (Hz): 0.848485\n"
-        "      time-hint-network (s): 5.6\n"
-        "      time-hint-ignore (s): 5.7\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      gpu-energy (J): 888\n"
-        "      gpu-power (W): 653\n"
-        "      gpu-frequency (Hz): 890\n"
-        "      uncore-frequency (Hz): 198\n"
-        "      CPU_ENERGY@package-0: 222\n"
-        "      CPU_ENERGY@package-1: 222\n"
-        "      agent stat: 2\n"
-        "    Unmarked Totals:\n"
-        "      runtime (s): 0.56\n"
-        "      count: 0\n"
-        "      sync-runtime (s): 444\n"
-        "      package-energy (J): 111\n"
-        "      dram-energy (J): 111\n"
-        "      power (W): 0.25\n"
-        "      frequency (%): 77.2727\n"
-        "      frequency (Hz): 0.772727\n"
-        "      time-hint-network (s): 1.2\n"
-        "      time-hint-ignore (s): 1.3\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      gpu-energy (J): 222\n"
-        "      gpu-power (W): 211\n"
-        "      gpu-frequency (Hz): 123\n"
-        "      uncore-frequency (Hz): 421\n"
-        "      CPU_ENERGY@package-0: 55.5\n"
-        "      CPU_ENERGY@package-1: 55.5\n"
-        "      agent stat: 3\n"
-        "    Epoch Totals:\n"
-        "      runtime (s): 70\n"
-        "      count: 66\n"
-        "      sync-runtime (s): 70\n"
-        "      package-energy (J): 167\n"
-        "      dram-energy (J): 167\n"
-        "      power (W): 2.38571\n"
-        "      frequency (%): 88.6364\n"
-        "      frequency (Hz): 0.886364\n"
-        "      time-hint-network (s): 4.2\n"
-        "      time-hint-ignore (s): 4.3\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      gpu-energy (J): 334\n"
-        "      gpu-power (W): 432\n"
-        "      gpu-frequency (Hz): 456\n"
-        "      uncore-frequency (Hz): 653\n"
-        "      CPU_ENERGY@package-0: 83.5\n"
-        "      CPU_ENERGY@package-1: 83.5\n"
-        "    Application Totals:\n"
-        "      runtime (s): 56\n"
-        "      count: 0\n"
-        "      sync-runtime (s): 56\n"
-        "      package-energy (J): 2222\n"
-        "      dram-energy (J): 2222\n"
-        "      power (W): 39.6786\n"
-        "      frequency (%): 66.6447\n"
-        "      frequency (Hz): 0.666447\n"
-        "      time-hint-network (s): 45\n"
-        "      time-hint-ignore (s): 46\n"
-        "      time-hint-compute (s): 0.2\n"
-        "      time-hint-memory (s): 0.3\n"
-        "      time-hint-io (s): 0.4\n"
-        "      time-hint-serial (s): 0.5\n"
-        "      time-hint-parallel (s): 0.6\n"
-        "      time-hint-unknown (s): 0.7\n"
-        "      time-hint-unset (s): 0.8\n"
-        "      time-hint-spin (s): 0.9\n"
-        "      gpu-energy (J): 4444\n"
-        "      gpu-power (W): 8992\n"
-        "      gpu-frequency (Hz): 74489\n"
-        "      uncore-frequency (Hz): 121213\n"
-        "      CPU_ENERGY@package-0: 1111\n"
-        "      CPU_ENERGY@package-1: 1111\n"
-        "      MPI startup (s): 22.11\n"
-        "      GEOPM startup (s): 0.321\n"
-        "      GEOPM overhead (s): 0.123\n"
-        "      geopmctl memory HWM (B): @ANY_STRING@\n"
-        "      geopmctl network BW (B/s): 678\n\n";
+    std::ostringstream expected;
+    expected << "GEOPM Version: " << std::string(geopm_version()) << "\n"
+             << "Start Time: " << m_start_time << "\n"
+             << "Profile: " << m_profile_name << "\n"
+             << "Agent: my_agent\n"
+             << "Policy: DYNAMIC\n"
+             << "one: 1\n"
+             << "two: 2\n"
+             << "\n"
+             << "Hosts:\n"
+             << "  " << geopm::hostname() << ":\n"
+             << "    three: 3\n"
+             << "    four: 4\n"
+             << "    Regions:\n"
+             << "    -\n"
+             << "      region: \"all2all\"\n"
+             << "      hash: 0x" << std::hex << std::setfill('0') << std::setw(8) << std::fixed << geopm_crc32_str("all2all") << "\n"
+             << "      runtime (s): 33.33\n"
+             << "      count: 20\n"
+             << "      sync-runtime (s): 555\n"
+             << "      package-energy (J): 388.5\n"
+             << "      dram-energy (J): 388.5\n"
+             << "      power (W): 0.7\n"
+             << "      frequency (%): 81.8182\n"
+             << "      frequency (Hz): 0.818182\n"
+             << "      time-hint-network (s): 3.4\n"
+             << "      time-hint-ignore (s): 3.5\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      gpu-energy (J): 777\n"
+             << "      gpu-power (W): 764\n"
+             << "      gpu-frequency (Hz): 567\n"
+             << "      uncore-frequency (Hz): 755\n"
+             << "      CPU_ENERGY@package-0: 194.25\n"
+             << "      CPU_ENERGY@package-1: 194.25\n"
+             << "      agent stat: 1\n"
+             << "      agent other stat: 2\n"
+             << "    -\n"
+             << "      region: \"MPI_Init_thread\"\n"
+             << "      hash: 0x" << std::hex << std::setfill('0') << std::setw(8) << std::fixed << geopm_crc32_str("MPI_Init_thread") << "\n"
+             << "      runtime (s): 22.11\n"
+             << "      count: 1\n"
+             << "      sync-runtime (s): 333\n"
+             << "      package-energy (J): 444\n"
+             << "      dram-energy (J): 444\n"
+             << "      power (W): 1.33333\n"
+             << "      frequency (%): 84.8485\n"
+             << "      frequency (Hz): 0.848485\n"
+             << "      time-hint-network (s): 5.6\n"
+             << "      time-hint-ignore (s): 5.7\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      gpu-energy (J): 888\n"
+             << "      gpu-power (W): 653\n"
+             << "      gpu-frequency (Hz): 890\n"
+             << "      uncore-frequency (Hz): 198\n"
+             << "      CPU_ENERGY@package-0: 222\n"
+             << "      CPU_ENERGY@package-1: 222\n"
+             << "      agent stat: 2\n"
+             << "    Unmarked Totals:\n"
+             << "      runtime (s): 0.56\n"
+             << "      count: 0\n"
+             << "      sync-runtime (s): 444\n"
+             << "      package-energy (J): 111\n"
+             << "      dram-energy (J): 111\n"
+             << "      power (W): 0.25\n"
+             << "      frequency (%): 77.2727\n"
+             << "      frequency (Hz): 0.772727\n"
+             << "      time-hint-network (s): 1.2\n"
+             << "      time-hint-ignore (s): 1.3\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      gpu-energy (J): 222\n"
+             << "      gpu-power (W): 211\n"
+             << "      gpu-frequency (Hz): 123\n"
+             << "      uncore-frequency (Hz): 421\n"
+             << "      CPU_ENERGY@package-0: 55.5\n"
+             << "      CPU_ENERGY@package-1: 55.5\n"
+             << "      agent stat: 3\n"
+             << "    Epoch Totals:\n"
+             << "      runtime (s): 70\n"
+             << "      count: 66\n"
+             << "      sync-runtime (s): 70\n"
+             << "      package-energy (J): 167\n"
+             << "      dram-energy (J): 167\n"
+             << "      power (W): 2.38571\n"
+             << "      frequency (%): 88.6364\n"
+             << "      frequency (Hz): 0.886364\n"
+             << "      time-hint-network (s): 4.2\n"
+             << "      time-hint-ignore (s): 4.3\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      gpu-energy (J): 334\n"
+             << "      gpu-power (W): 432\n"
+             << "      gpu-frequency (Hz): 456\n"
+             << "      uncore-frequency (Hz): 653\n"
+             << "      CPU_ENERGY@package-0: 83.5\n"
+             << "      CPU_ENERGY@package-1: 83.5\n"
+             << "    Application Totals:\n"
+             << "      runtime (s): 56\n"
+             << "      count: 0\n"
+             << "      sync-runtime (s): 56\n"
+             << "      package-energy (J): 2222\n"
+             << "      dram-energy (J): 2222\n"
+             << "      power (W): 39.6786\n"
+             << "      frequency (%): 66.6447\n"
+             << "      frequency (Hz): 0.666447\n"
+             << "      time-hint-network (s): 45\n"
+             << "      time-hint-ignore (s): 46\n"
+             << "      time-hint-compute (s): 0.2\n"
+             << "      time-hint-memory (s): 0.3\n"
+             << "      time-hint-io (s): 0.4\n"
+             << "      time-hint-serial (s): 0.5\n"
+             << "      time-hint-parallel (s): 0.6\n"
+             << "      time-hint-unknown (s): 0.7\n"
+             << "      time-hint-unset (s): 0.8\n"
+             << "      time-hint-spin (s): 0.9\n"
+             << "      gpu-energy (J): 4444\n"
+             << "      gpu-power (W): 8992\n"
+             << "      gpu-frequency (Hz): 74489\n"
+             << "      uncore-frequency (Hz): 121213\n"
+             << "      CPU_ENERGY@package-0: 1111\n"
+             << "      CPU_ENERGY@package-1: 1111\n"
+             << "      MPI startup (s): 22.11\n"
+             << "      GEOPM startup (s): 0.321\n"
+             << "      GEOPM overhead (s): 0.123\n"
+             << "      geopmctl memory HWM (B): @ANY_STRING@\n"
+             << "      geopmctl network BW (B/s): 678\n\n";
 
-    std::istringstream exp_stream(expected);
-
+    std::istringstream exp_istream(expected.str());
     m_reporter->update();
     m_reporter->overhead(0.123, 0.321);
     m_reporter->generate("my_agent", agent_header, agent_node_report, m_region_agent_detail,
                          m_application_io,
                          m_comm, m_tree_comm);
     std::ifstream report(m_report_name);
-    check_report(exp_stream, report);
+    check_report(exp_istream, report);
 }
 
 void check_report(std::istream &expected, std::istream &result)

